mythril/support/signatures.py

```diff
@@ -1,44 +1,244 @@
+
+        self.locked = False
+
+
+
+class SignatureDb(object):
+
+    def __init__(self, enable_online_lookup=True):
+        """
+        Constr
+        :param enable_online_lookup: enable onlien signature hash lookup
+        """
+        self.signatures = {}  # signatures in-mem cache
+        self.signatures_file = None
+        self.signatures_file_lock = None
+        self.enable_online_lookup = enable_online_lookup  # enable online funcsig resolving
+        self.online_lookup_miss = set()  # temporarily track misses from onlinedb to avoid requesting the same non-existent sighash multiple times
+        self.online_directory_unavailable_until = 0  # flag the online directory as unavailable for some time
+
+    def open(self, path=None):
+        """
+        Open a function signature db from json file
+
+        :param path: specific path to signatures.json; default mythril location if not specified
+        :return: self
+        """
+        if not path:
+            #  try default locations
+            try:
+                mythril_dir = os.environ['MYTHRIL_DIR']
+            except KeyError:
+                mythril_dir = os.path.join(os.path.expanduser('~'), ".mythril")
+            path = os.path.join(mythril_dir, 'signatures.json')
+
+        self.signatures_file = path  # store early to allow error handling to access the place we tried to load the file
+
+        if not os.path.exists(path):
+            logging.debug("Signatures: file not found: %s" % path)
+            raise FileNotFoundError("Missing function signature file. Resolving of function names disabled.")
+
+        self.signatures_file_lock = SimpleFileLock(self.signatures_file)  # lock file to prevent concurrency issues
+        self.signatures_file_lock.aquire()  # try to aquire it within the next 10s
+
+        with open(path, 'r') as f:
+            sigs = json.load(f)
+
+        self.signatures_file_lock.release()  # release lock
+
+        # normalize it to {sighash:list(signatures,...)}
+        for sighash, funcsig in sigs.items():
+            if isinstance(funcsig, list):
+                self.signatures = sigs  # keep original todo: tintinweb - super hacky. make sure signatures.json is initially in correct format fixme
+                break  # already normalized
+            self.signatures.setdefault(sighash, [])
+            self.signatures[sighash].append(funcsig)
+
+        return self
+
+    def write(self, path=None, sync=True):
+        """
+        Write signatures database as json to file
+
+        :param path: specify path otherwise update the file that was loaded with open()
+        :param sync: lock signature file, load contents and merge it into memcached sighash db, then save it
+        :return: self
+        """
+        path = path or self.signatures_file
+        self.signatures_file_lock = SimpleFileLock(path)  # lock file to prevent concurrency issues
```

Comment:
maybe only initialize if signatures_file_lock is None

Reply:
shouldn't make much of a difference besides GCing the old lock object and creating a new instance. Since `db.write()` can be called without ever calling `db.open()` I'll change this to use an existing instance if available.